packages/test/test-end-to-end-tests/src/test/dataMigration.spec.ts

```diff
@@ -0,0 +1,425 @@
+		this._getter = getter;
+	}
+
+	protected async initializingFromExisting(): Promise<void> {
+		const handle = this.root.get<IFluidHandle<SharedMap>>(getterKey);
+		this._getter = await handle?.get();
+	}
+}
+
+function createReadOnlyContext(
+	context: IContainerContext,
+	codeDetailsVersion: string,
+): IContainerContext {
+	const readonlyContext: IContainerContext = {
+		...context,
+		connected: true,
+		clientDetails: {
+			capabilities: context.clientDetails.capabilities,
+			type: detachedClientType,
+		},
+		storage: {
+			...context.storage,
+			getSnapshotTree: context.storage.getSnapshotTree,
+			getVersions: context.storage.getVersions,
+			createBlob: async (_file: ArrayBufferLike) => {
+				throw new Error("Should not be submitting blobs to storage");
+			},
+			readBlob: context.storage.readBlob,
+			uploadSummaryWithContext: async (_summary, _context) => {
+				throw new Error("Should not be uploading a summary");
+			},
+			downloadSummary: context.storage.downloadSummary,
+		},
+		submitFn: (_message, _contents, _batch, _appData) => 0,
+		submitSummaryFn: (_summaryOp, _referenceSequenceNumber?) => 0,
+		submitBatchFn: (_batch, _referenceSequenceNumber?) => 0,
+		submitSignalFn: (_contents: any) => {},
+		disposeFn: context.disposeFn,
+		closeFn: context.closeFn,
+		updateDirtyContainerState: context.updateDirtyContainerState,
+		getAbsoluteUrl: context.getAbsoluteUrl,
+		getSpecifiedCodeDetails: () => {
+			return { package: codeDetailsVersion };
+		},
+		getLoadedFromVersion: context.getLoadedFromVersion,
+		attachState: context.attachState,
+	};
+	return readonlyContext;
+}
+
+class ContainerRuntimeFactoryWithDataMigration extends RuntimeFactoryHelper {
+	private _currentRuntimeFactory?: ContainerRuntimeFactoryWithDefaultDataStore;
+	private oldRuntime?: ContainerRuntime;
+	private get currentRuntimeFactory(): ContainerRuntimeFactoryWithDefaultDataStore {
+		assert(
+			this._currentRuntimeFactory !== undefined,
+			"current should have been set before getting!",
+		);
+		return this._currentRuntimeFactory;
+	}
+
+	private clientDetailsType?: string;
+
+	constructor(
+		private readonly v1: ContainerRuntimeFactoryWithDefaultDataStore,
+		private readonly v1ToV2: ContainerRuntimeFactoryWithDefaultDataStore,
+		private readonly v2: ContainerRuntimeFactoryWithDefaultDataStore,
+		private readonly deferred: Deferred<{
+			summarizer: ContainerRuntime;
+			readonlyRuntime: ContainerRuntime;
+		}>,
+	) {
+		super();
+	}
+	public async preInitialize(
+		context: IContainerContext,
+		existing: boolean,
+	): Promise<ContainerRuntime> {
+		this.clientDetailsType = context.clientDetails.type;
+		const codeDetails = context.getSpecifiedCodeDetails
+			? context.getSpecifiedCodeDetails()
+			: undefined;
+		assert(codeDetails !== undefined, "get code details failed");
+
+		this._currentRuntimeFactory = codeDetails.package === "v1" ? this.v1 : this.v2;
+		if (this.clientDetailsType === detachedClientType) {
```

Comment:
Something worth confirming (I think I made this mistake initially in my thinking, and I think you follow it):
I think CRF is supposed to be stateless object, and used to load many different files (i.e. there is one-many relationships for CFR-CR). If that's true, then I think we are in trouble with this approach.

If you look at IRuntimeFactory, it's simple - it has only one method. I think using RuntimeFactoryHelper here hurts more than it helps - if it's a single method, then that means all this code could be written in linear fashion, and then most of the state likely will be on stack, not as class remembers.

That said, I think the whole RuntimeFactoryHelper pattern is wrong. I'd break it into two classes (probably worth opening a task to do this work in the future, as it's outside of this work):
1. Class A, that implements IRuntimeFactory, but also takes a factory to be able to create instances of class B for each instantiation.
2. Class B, that has all the methods around pre-initializaiton, post-initializaiton, etc. And it's stateful, i.e. it's safe to have state because it's used to load only one instance of a runtime.
3. I'd kill "provide" nonsense from it, and make it not implement IRuntimeFactory, but implement all the other methods it has today. And in places where we need IRuntimeFactory, I'd use different class implementation that would create

Reply:
I'm not in love with this implementation either. Glad this clarified that part.